src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,572 @@
+        }
+        if nodes.is_empty() {
+            return Err(InvalidSubgraph::EmptySubgraph);
+        }
+        let Some(parent) = base.get_parent(nodes[0]) else {
+            return Err(InvalidSubgraph::NoSharedParent);
+        };
+        if !nodes.iter().all(|&n| base.get_parent(n) == Some(parent)) {
+            return Err(InvalidSubgraph::NoSharedParent);
+        }
+        Ok(Self { base, nodes })
+    }
+
+    /// An iterator over the nodes in the subgraph.
+    pub fn nodes(&self) -> &[Node] {
+        &self.nodes
+    }
+
+    /// Whether a port is at the subgraph boundary.
+    fn is_boundary_port(&self, n: Node, p: Port) -> bool
+    where
+        Base: HugrView,
+    {
+        self.base
+            .linked_ports(n, p)
+            .any(|(n, _)| !self.nodes.contains(&n))
+    }
+
+    /// An iterator of the incoming boundary ports.
+    fn incoming_ports(&self) -> impl Iterator<Item = (Node, Port)> + '_
+    where
+        Base: HugrView,
+    {
+        self.boundary_ports(Direction::Incoming)
+    }
+
+    /// An iterator of the outgoing boundary ports.
+    fn outgoing_ports(&self) -> impl Iterator<Item = (Node, Port)> + '_
+    where
+        Base: HugrView,
+    {
+        self.boundary_ports(Direction::Outgoing)
+    }
+
+    /// An iterator of the boundary ports, either incoming or outgoing.
+    fn boundary_ports(&self, dir: Direction) -> impl Iterator<Item = (Node, Port)> + '_
+    where
+        Base: HugrView,
+    {
+        self.nodes.iter().flat_map(move |&n| {
+            self.base
+                .node_ports(n, dir)
+                .filter(move |&p| self.is_boundary_port(n, p))
+                .map(move |p| (n, p))
+        })
+    }
+
+    /// The signature of the subgraph.
+    pub fn signature(&self) -> AbstractSignature
+    where
+        Base: HugrView,
+    {
+        let (input, input_static): (Vec<_>, Vec<_>) = self
+            .incoming_ports()
+            .filter_map(|(n, p)| self.base.get_optype(n).signature().get(p))
+            .partition_map(|edgekind| match edgekind {
+                EdgeKind::Value(t) => Either::Left(t),
+                EdgeKind::Static(t) => Either::Right(t),
+                _ => unimplemented!("ControlFlow and StateOrder edges not supported"),
+            });
+        let output = self
+            .outgoing_ports()
+            .filter_map(|(n, p)| self.base.get_optype(n).signature().get(p))
+            .map(|edgekind| {
+                if let EdgeKind::Value(t) = edgekind {
+                    t
+                } else {
+                    unimplemented!(
+                        "ControlFlow, StateOrder and Static edges not supported for output"
+                    )
+                }
+            })
+            .collect_vec();
+        AbstractSignature::new(input, output, input_static)
+    }
+
+    /// The parent of the sibling subgraph.
+    pub fn get_parent(&self) -> Node
+    where
+        Base: HugrView,
+    {
+        self.base
```

Comment:
Don't we have `self.base.root()`?

Reply:
A subgraph could have any node as its parent. As currently implemented, this is not necessarily the root of the Hugr view.

Reply:
Oh I see, I misunderstood you. The problem is that `self.base` might not be a SiblingGraph, it could be a generic HUGR.

What is true is that `self.get_parent() == self.root()` *IF* `SiblingSubgraph` was implementing `HugrView`. At the moment it is not. I've added this as a future TODO for another PR.

What do you suggest I do?

Reply:
What I meant was:
```
fn get_parent(&self) -> Node {
  self.base.root()
}
```

Reply:
Hmmmm, ok. I could argue that `SiblingGraph` is the thing that selects a group of siblings, and then this should be a subset of that - so you should embed a `SiblingGraph` struct directly inside this. But I can see that that'd be annoying as much as anything :-(.... so ok, yeah. I don't see anything better really

Reply:
(but ok, fine as it is)